feat: patch connections created by a pool's getConnection

M-TGH opened the pull request:
*Issue #, if available:* #87

*Description of changes:*

Adds a simple patchObject call around connections returned from a pools `getConnection`.
Far as I could tell mysql doesn't take any other arguments for this, so it assumes the first argument is the callback.

I'm unsure what mysql2 makes of this, but I assume they promisified it which I didn't yet support. Would be nice to hear from someone who uses mysql2 what changes are still needed to support mysql2. I'll be looking into adding it but would like confirmation on what parameters it takes and what it returns.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

M-TGH commented:
Added what I think should be all that's needed for mysql2 support, still wouldn't mind if someone who's more familiar with mysql2 could take a look at it as well.

darcoli commented:
Tried using this within our application which uses `typeORM` with the `mysql` driver, however I kept getting the following exception:
```
'RangeError: Maximum call stack size exceeded',
'    at Subsegment.init (/var/task/dist/webpack:/application/node_modules/aws-xray-sdk-core/lib/segments/attributes/subsegment.js:32:1)',
'    at new Subsegment (/var/task/dist/webpack:/application/node_modules/aws-xray-sdk-core/lib/segments/attributes/subsegment.js:25:1)',
'    at Segment.addNewSubsegment (/var/task/dist/webpack:/application/node_modules/aws-xray-sdk-core/lib/segments/segment.js:179:1)',
'    at PoolConnection.__query (/var/task/dist/webpack:/application/node_modules/aws-xray-sdk-mysql/lib/mysql_p.js:152:1)',
'    at PoolConnection.__query (/var/task/dist/webpack:/application/node_modules/aws-xray-sdk-mysql/lib/mysql_p.js:178:1)',
'    at PoolConnection.__query (/var/task/dist/webpack:/application/node_modules/aws-xray-sdk-mysql/lib/mysql_p.js:178:1)',
'    at PoolConnection.__query (/var/task/dist/webpack:/application/node_modules/aws-xray-sdk-mysql/lib/mysql_p.js:178:1)',
'    at PoolConnection.__query (/var/task/dist/webpack:/application/node_modules/aws-xray-sdk-mysql/lib/mysql_p.js:178:1)',
'    at PoolConnection.__query (/var/task/dist/webpack:/application/node_modules/aws-xray-sdk-mysql/lib/mysql_p.js:178:1)' ] } } }   
```

I did get it to work by adding further conditions to avoid the infinite recursion - changing the function `patchObject()` as follows:
```
function patchObject(connection) {
  if (connection.query instanceof Function) {
    if (!connection.__query) {
      connection.__query = connection.query;
      connection.query = captureOperation('query');
    }
  } 
    
  if (connection.execute instanceof Function) {
    if (!connection.__execute) {
      connection.__execute = connection.execute;
      connection.execute = captureOperation('execute');
    }
  }
      
  if (connection.getConnection instanceof Function) { 
    if (!(connection['__getConnection']) ) {
      patchGetConnection(connection);
    }
  }
  return connection;
}
```

M-TGH commented:
Thanks @darcoli, I've added your feedback in e0fadd4.

darcoli commented:
@M-TGH thanks for adding the changes - did not get any issues so far ;)

Any chance of getting this merged any time soon?

chrisradek commented on packages/mysql/lib/mysql_p.js:
```diff
@@ -66,16 +66,42 @@ function patchCreatePool(mysql) {
   };
 }
 
+function patchGetConnection(pool) {
```
Would this also work with `PoolCluster` if instead of grabbing the callback from the 1st argument, you grabbed it from the last argument?

https://github.com/mysqljs/mysql#poolcluster

M-TGH replied:
Looks like it would work, I'll add it into this PR, if you prefer me to split it up into a separate PR let me know (it may make the PR a bit big in size).

M-TGH commented:
@chrisradek I've added basic support for mysql.createPoolCluster in dfc1c55 (getConnection & query). Let me know if you would prefer it in a separate PR.

I've also added support for poolCluster.of in db8e782, however I'm on the fence about patching of this way since it's rather ambiguous.

M-TGH commented:
Good catch on the args, added a fix in 5bb11c8.

chrisradek commented on packages/mysql/lib/mysql_p.js:
```diff
@@ -66,16 +67,72 @@ function patchCreatePool(mysql) {
+    }
+
+    var result = pool[baseFcn].apply(pool, args);
+    if (result instanceof Promise) return result.then(patchObject);
```
Sorry I missed this before, but could you instead check if `result` has a method `then`? I've seen this check bite us in other libraries when a user has replaced the native `Promise` class with a different once, e.g. bluebird.

M-TGH commented:
Updated the check in 1937797.
Whilst replacing this `instanceof Promise` check I noticed there's 2 more:
https://github.com/aws/aws-xray-sdk-node/blob/a8741c0151fbc43df621afd50229616533687d80/packages/mysql/lib/mysql_p.js#L39
https://github.com/aws/aws-xray-sdk-node/blob/a8741c0151fbc43df621afd50229616533687d80/packages/mysql/lib/mysql_p.js#L57

I'll replace those in a separate PR since they weren't added in this one, however let me know if you prefer it included in this PR.
